### PR TITLE
First draft of 1.34 announcement

### PR DESCRIPTION
**Pull request by Centril**
r? @rust-lang/release @rust-lang/core

Written in collaboration with @Manishearth, @ehuss, and @Mark-Simulacrum and proof read by others.

### CODE REVIEW

File: posts/2019-04-11-Rust-1.34.0.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+/// let mut input = String::new();
+/// io::stdin().read_line(&mut input)?;
+/// # Ok::<(), io:Error>(())
+/// ```
+fn my_func() {}
+````
+
+You still have to specify the error type being used at the bottom of the documentation test.
+
+### Custom attributes accept arbitrary token streams
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]: r? @petrochenkov

File: posts/2019-04-11-Rust-1.34.0.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+```rust
+let num = u32::from_be_bytes(slice.try_into()?);
+```
+
+For conversions that can't fail, such as `u8` to `u32`, the [`Infallible`] type was added.
+This also permits a blanket implementation of `TryFrom` for all existing `From` implementations.
+In the future, we hope to turn `Infallible` into an alias for [the `!` (never) type][never_type].
+
+### `fn before_exec` deprecated in favor of `unsafe fn pre_exec`
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]: r? @RalfJung

Comment (RalfJung) [reply posts/2019-04-11-Rust-1.34.0.md]: Might be worth mentioning the other concerns related to signal-safety in forked code?

> After a fork() in a multithreaded program, the child can safely call only async-signal-safe functions (see signal-safety(7)) until such time as it calls execve(2).

(From the fork man page)

For example, acquiring a lock in `pre_exec` is a very bad idea because if the lock was held by *someone* when the fork happened, now you got a deadlock.

Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: Sure, not sure how to phrase it. Could you file a PR against my branch or use GH suggestions?

File: posts/2019-04-11-Rust-1.34.0.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+```rust
+#[foo(bar)]
+#[foo = "bar"]
+#[foo = 0]
+#[foo(bar = true)]
+#[foo(bar, baz(quux, foo = "bar"))]
+```
+
+Unlike `macro_rules!`, these attributes could not accept arbitrary token streams in delimiters,
```
Comment (petrochenkov) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
Unlike procedural macros themselves, these helper attributes could not accept arbitrary token streams in delimiters,
```
`macro_rules!` seems mostly unrelated here.

File: posts/2019-04-11-Rust-1.34.0.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+in the `[dependencies]` section of your `Cargo.toml`.
+
+Not everyone _wants_ to publish their crates to crates.io, however.
+People maintaining proprietary/closed-source code can't use crates.io,
+and instead are forced to use `git` or `path` dependencies.
+This is usually fine for small projects, but if you have a lot of closed-source crates
+within a large organization, you lose the benefit of the versioning support that crates.io has.
+
+With this release, Cargo gains support for alternate registries.
+These registries coexist with crates.io, so you can write software that depends
```
Comment (pietroalbini) [inline posts/2019-04-11-Rust-1.34.0.md]: We might want to specify that crates on crates.io can't depend on external registries.

**Comment by Manishearth** [comment]
"unlike `foo!()`-style macros" or "bang macros" might make more sense, that
was the intent of this text

**Comment by Centril** [comment]
@Manishearth On re-reading the stabilization PR, it says:

> Delimited non-macro non-builtin attributes now support the same syntax as macro attributes:

so I think @petrochenkov's change was right (and he wrote the PR :D).

File: posts/2019-04-11-Rust-1.34.0.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+#[foo = 0]
+#[foo(bar = true)]
+#[foo(bar, baz(quux, foo = "bar"))]
+```
+
+Unlike procedural macros themselves, these helper attributes could not accept arbitrary token streams in delimiters,
+so you could not write `#[range(0..10)]` or `#[bound(T: MyTrait)]`.
+Procedural macro crates would instead use strings for specifying syntaxes like this, e.g. `#[range("0..10")]`
+
+With this Rust release, custom attributes `#[attr($tokens)]` now accept
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
With this Rust release, custom attributes `#[attr($tokens)]` [now accept][arbitrary-tts]
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+/// # Ok::<(), io:Error>(())
+/// ```
+fn my_func() {}
+````
+
+You still have to specify the error type being used at the bottom of the documentation test.
+
+### Custom attributes accept arbitrary token streams
+
+[Procedural macros]: https://blog.rust-lang.org/2018/12/21/Procedural-Macros-in-Rust-2018.html
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
[Procedural macros]: https://blog.rust-lang.org/2018/12/21/Procedural-Macros-in-Rust-2018.html
[arbitrary-tts]: https://github.com/rust-lang/rust/pull/57367
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+Not everyone _wants_ to publish their crates to crates.io, however.
+People maintaining proprietary/closed-source code can't use crates.io,
+and instead are forced to use `git` or `path` dependencies.
+This is usually fine for small projects, but if you have a lot of closed-source crates
+within a large organization, you lose the benefit of the versioning support that crates.io has.
+
+With this release, Cargo gains support for alternate registries.
+These registries coexist with crates.io, so you can write software that depends
+on crates from both crates.io and your custom registry.
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
on crates from both crates.io and your custom registry.
Crates on crates.io cannot however depend on external registries.
```

@pietroalbini like this?

File: posts/2019-04-11-Rust-1.34.0.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+[crates.io]: http://crates.io/
+[registry-docs]: https://doc.rust-lang.org/nightly/cargo/reference/registries.html#running-a-registry
+
+Since before 1.0, Rust has had a public crate registry, [crates.io].
+People publish crates with `cargo publish` and it's easy to include these crates
+in the `[dependencies]` section of your `Cargo.toml`.
+
+Not everyone _wants_ to publish their crates to crates.io, however.
+People maintaining proprietary/closed-source code can't use crates.io,
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
People maintaining proprietary/closed-source code cannot use crates.io,
```

File: posts/2019-04-11-Rust-1.34.0.md (hunk 7 above)
Comment (pietroalbini) [reply posts/2019-04-11-Rust-1.34.0.md]: Yeah, looks good :+1:

File: posts/2019-04-11-Rust-1.34.0.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+For example, the [`from_be_bytes`] and related methods on integer types take arrays,
+but data is often read in via slices. Converting between slices and arrays is tedious to do manually.
+With the new traits, it can be done inline with `.try_into()`.
+
+```rust
+let num = u32::from_be_bytes(slice.try_into()?);
+```
+
+For conversions that can't fail, such as `u8` to `u32`, the [`Infallible`] type was added.
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
For conversions that cannot fail, such as `u8` to `u32`, the [`Infallible`] type was added.
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+### Alternative `cargo` registries
+
+[crates.io]: http://crates.io/
+[registry-docs]: https://doc.rust-lang.org/nightly/cargo/reference/registries.html#running-a-registry
+
+Since before 1.0, Rust has had a public crate registry, [crates.io].
+People publish crates with `cargo publish` and it's easy to include these crates
+in the `[dependencies]` section of your `Cargo.toml`.
+
+Not everyone _wants_ to publish their crates to crates.io, however.
```
Comment (Dylan-DPC) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
However, not everyone _wants_ to publish their crates to crates.io.
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+```rust
+#[foo(bar)]
+#[foo = "bar"]
+#[foo = 0]
+#[foo(bar = true)]
+#[foo(bar, baz(quux, foo = "bar"))]
+```
+
+Unlike procedural macros themselves, these helper attributes could not accept arbitrary token streams in delimiters,
```
Comment (Dylan-DPC) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
Unlike procedural macros, these helper attributes could not accept arbitrary token streams in delimiters,
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+[`NonZeroU8`]: https://doc.rust-lang.org/std/num/struct.NonZeroU8.html
+[`NonZeroI8`]: https://doc.rust-lang.org/std/num/struct.NonZeroI8.html
+[`iter::from_fn`]: https://doc.rust-lang.org/std/iter/fn.from_fn.html
+[`iter::successors`]: https://doc.rust-lang.org/std/iter/fn.successors.html
+[prev-1.28]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1280-2018-08-02
+
+In 1.34.0, the set of stable atomic integer types was expanded,
+with signed and unsigned variants from 8 ([`AtomicU8`]) to 64 bits now available.
+
+[Previously][prev-1.28], non-zero unsigned integer types, e.g. [`NonZeroU8`], was stabilized.
```
Comment (jonas-schievink) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
[Previously][prev-1.28], non-zero unsigned integer types, e.g. [`NonZeroU8`], were stabilized.
```

File: posts/2019-04-11-Rust-1.34.0.md [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+```sh
+cargo publish --registry=my-registry
+```
+
+There is [documentation][registry-docs] on how to run your own registry.
+
+### `?` in documentation tests
+
+[RFC 1937]: http://rust-lang.github.io/rfcs/1937-ques-in-main.html
```
Comment (jonas-schievink) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
[RFC 1937]: https://rust-lang.github.io/rfcs/1937-ques-in-main.html
```

File: posts/2019-04-11-Rust-1.34.0.md (hunk 2 above)
Comment (RalfJung) [reply posts/2019-04-11-Rust-1.34.0.md]: I'm afraid I don't have any good ideas either, and this POSIX stuff is not at all my expertise.

Comment (alexcrichton) [reply posts/2019-04-11-Rust-1.34.0.md]: [According to what the closest thing I know of to the POSIX spec is](http://pubs.opengroup.org/onlinepubs/007904975/functions/fork.html) it's similar to what @RalfJung quoted but subtly different:

> Consequently, to avoid errors, the child process may only execute async-signal-safe operations until such time as one of the exec functions is called

My interpretation that jives with my understanding of the API, which is that it's tough to use `before_exec` correctly but from the perspective of POSIX it's no UB, but rather we're explicitly saying it's UB with respect to file descriptor duplication (and how `FromRawFd` has various guarantees and whatnot).

I don't think we settled on any other concrete sequence of actions we'd consider UB other than fd trickery

Comment (RalfJung) [reply posts/2019-04-11-Rust-1.34.0.md]: What I quoted is from `man fork` on my system, also [available online](https://www.systutorials.com/docs/linux/man/2-fork/). [This page](http://man7.org/linux/man-pages/man7/signal-safety.7.html) talks about "unpredictable results" and things being "unsafe to call" when in a not-async-signal-safe function in a signal handler; I do not know if that means UB or something weaker. I see no reason why calling them from a forked-off process should be any different.

Indeed we have no other UB that I know of, but I also did not try to construct a Rust library that is generally safe to use but UB when called in a not-async-signal-safe context. But we certainly have a chance of deadlocks (like when you try to acquire a mutex), and a case of nobody-knows-what-happens when you call into C code that does not declare that it is async-signal-safe.

Anyway my intention was not to say here that these things are UB, but to call out the fact that this is a constrained environment and very unpredictable but possibly not UB things can certainly happen.

**Comment by steveklabnik** [comment]
This looks good to me, however I think we should remove the thanks link. I can't guarantee I'll have it up and running before the release goes out, though it should be able to work later today. @Mark-Simulacrum and I are gonna chat about how to not block this one me so much, but for now, it's probably best.

**Comment by Mark-Simulacrum** [comment]
Removed the thanks link.

File: posts/2019-04-11-Rust-1.34.0.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
+---
+layout: post
+title: "Announcing Rust 1.34.0"
+author: The Rust Release Team
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.34.0. Rust is a
+programming language that is empowering everyone to build reliable and
```
Comment (nathany) [inline posts/2019-04-11-Rust-1.34.0.md]: "that is empowering" could be more concisely "that empowers"

File: posts/2019-04-11-Rust-1.34.0.md [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+
+If you have a previous version of Rust installed via rustup, getting Rust
+1.34.0 is as easy as:
+
+```console
+$ rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install] from the appropriate
+page on our website, and check out the [detailed release notes for 1.34.0][notes] on GitHub.
```
Comment (nathany) [inline posts/2019-04-11-Rust-1.34.0.md]: The way this is written, it suggests that (only) people who don't have rust(up) already should check out the release notes. 

The opposite is probably true -- new people need not check out the release notes, people already familiar with Rust will benefit more from doing so.

Since the release notes are mentioned later, I'd suggest removing that part of the sentence here.

File: posts/2019-04-11-Rust-1.34.0.md [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+[`from_be_bytes`]: https://doc.rust-lang.org/std/primitive.u32.html#method.from_be_bytes
+[never_type]: https://github.com/rust-lang/rust/issues/35121
+[`TryFrom`]: https://doc.rust-lang.org/std/convert/trait.TryFrom.html
+[`TryInto`]: https://doc.rust-lang.org/std/convert/trait.TryInto.html
+[`Infallible`]: https://doc.rust-lang.org/std/convert/enum.Infallible.html
+
+The [`TryFrom`] and [`TryInto`] traits were stabilized to allow fallible type conversions.
+
+For example, the [`from_be_bytes`] and related methods on integer types take arrays,
+but data is often read in via slices. Converting between slices and arrays is tedious to do manually.
```
Comment (nathany) [inline posts/2019-04-11-Rust-1.34.0.md]: suggestion: ~~is~~ has been tedious

it no longer is 😃

File: posts/2019-04-11-Rust-1.34.0.md [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+[`CommandExt::before_exec`]: https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.before_exec
+[`CommandExt::pre_exec`]: https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.pre_exec
+[ub-possible]: https://github.com/rust-lang/rust/issues/39575#issuecomment-437658766
+[non-dup]: https://github.com/rust-lang/rust/issues/39575#issuecomment-439645949
+
+On Unix-like systems, the function [`CommandExt::before_exec`] allows you to
+schedule a closure to be run before `exec` is invoked.
+
+As the closure provided will be run in the context of the child process after a fork,
+this means that resources, such as file descriptors and memory-mapped regions, may get duplicated.
```
Comment (nathany) [inline posts/2019-04-11-Rust-1.34.0.md]: This sentence doesn't flow very well (try reading it aloud). Can it be reworded?

File: posts/2019-04-11-Rust-1.34.0.md (hunk 14 above)
Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: I think this is due to the website slogan:
> Empowering everyone to build reliable and efficient software. 

Maybe it should change, but probably not here and now =P

File: posts/2019-04-11-Rust-1.34.0.md (hunk 16 above)
Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: It is still tedious to do manually ;)

File: posts/2019-04-11-Rust-1.34.0.md (hunk 15 above)
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
page on our website.
```

File: posts/2019-04-11-Rust-1.34.0.md (hunk 14 above)
Comment (nathany) [reply posts/2019-04-11-Rust-1.34.0.md]: Maybe the slogan could be in quotes or italics here?

Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: Not sure, it's worth considering for the next release, but we are quite close to deadline now.

Comment (nathany) [reply posts/2019-04-11-Rust-1.34.0.md]: 👌

File: posts/2019-04-11-Rust-1.34.0.md [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+
+[`CommandExt::before_exec`]: https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.before_exec
+[`CommandExt::pre_exec`]: https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.pre_exec
+[ub-possible]: https://github.com/rust-lang/rust/issues/39575#issuecomment-437658766
+[non-dup]: https://github.com/rust-lang/rust/issues/39575#issuecomment-439645949
+
+On Unix-like systems, the function [`CommandExt::before_exec`] allows you to
+schedule a closure to be run before `exec` is invoked.
+
+As the closure provided will be run in the context of the child process after a fork,
```
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
The closure provided will be run in the context of the child process after a fork.
```

File: posts/2019-04-11-Rust-1.34.0.md (hunk 17 above)
Comment (Centril) [inline posts/2019-04-11-Rust-1.34.0.md]:
```suggestion
This means that resources, such as file descriptors and memory-mapped regions, may get duplicated.
```

Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: Changed ever so slightly :)

File: posts/2019-04-11-Rust-1.34.0.md (hunk 2 above)
Comment (Centril) [reply posts/2019-04-11-Rust-1.34.0.md]: So I think we don't have time to add or change anything here unfortunately... Maybe this can be improved in the libstd docs later...

File: posts/2019-04-11-Rust-1.34.0.md (hunk 17 above)
Comment (nathany) [reply posts/2019-04-11-Rust-1.34.0.md]: Thank you.
